include/deal.II/lac/utilities.h

```diff
@@ -0,0 +1,301 @@
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/lac/lapack_templates.h>
+#include <deal.II/lac/lapack_support.h>
+#include <deal.II/lac/vector_memory.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  /**
+   * A collection of linear-algebra utilities.
+   */
+  namespace LinearAlgebra
+  {
+
+    /**
+     * Estimate an upper bound for the largest eigenvalue of @p op by a @p k -step
+     * Lanczos process starting from the initial vector @p v0. Typical
+     * values of @p k are below 10. The returned
+     * value is $ ||T||_2 + ||f||_2$, where $H V_k=V_k T_k+f_k e_k^T$ and
+     * T is the the tridiagonal Lanczos matrix. If @p eigenvalues is
+     * not <code>nullptr</code>, the eigenvaleus of $T$ will be written there.
+     *
+     * @p vector_memory is used to allocate memory for temporary vectors.
+     * OperatorType has to provide <code>vmult</code> operation with
+     * VectorType.
+     *
+     * @note This function uses Lapack routines to compute the largest
+     * eigenvalue of T.
+     *
+     * @note It is an alternative estimate to that, obtained from
+     * several steps of SolverCG with SolverCG<VectorType>::connect_eigenvalues_slot().
+     *
+     * @author Denis Davydov, 2017
+     */
+    template <typename OperatorType, typename VectorType>
+    double Lanczos_largest_eigenvalue(const OperatorType &op,
+                                      const VectorType &v0,
+                                      const unsigned int k,
+                                      VectorMemory<VectorType> &vector_memory,
+                                      std::vector<double> *eigenvalues = nullptr);
+
+    /**
+     * Apply Chebyshev polynome of the operator @p H to @p x. For a
+     * non-defective operator with a complete set of eigenpairs
+     * $H \psi_i = \lambda_i \psi_i$, the action is given by
+     * $p(H)x = a_i p(\lambda_i) \psi_i$, where $x=: a_i \psi_i$. Thus
+     * by appropriately choosing the polynomial filter, one can alter
+     * the eigenmodes contained in $x$.
+     *
+     * This function uses Chebyshev polynomes of first kind. Below is an
+     * example of polynome of degree 8 normalized to unity at $-1.2$.
+     * <table>
+     *   <tr>
+     *     <td align="center">
+     *       @image html chebyshev8.png
+     *     </td>
+     *   </tr>
+     * </table>
+     * By introducing a linear mapping from @p unwanted_spectrum to
+     * $[-1,1]$, we can dump the corresponding modes in @p x. The higher
+     * the polynomial degree, the more rapid it grows outside of the
+     * $[-1,1]$. In order to avoid numerical overflow, we normalize
+     * polynomial filter to unity at @p tau. The action of the Chebyshev filter only requires
+     * evaluation of <code>vmult()</code> of @p H and is based on the
+     * recursion equation for Chebyshev polynomial of degree $n$:
+     * $T_{n}(x) = 2x T_{n-1}(x) - T_{n-2}(x)$ with $T_0(x)=1$ and $T_1(x)=x$.
```

Comment:
p is any polynomial, whereas T is Chebyshev polynomial. I added `$T_8(x)$` above to make it clear.